In [6]:
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:90% !important; }</style>"))

In [1]:
from IPython.display import SVG, Math
from fractions import Fraction as F
import matplotlib.pyplot as plt
import random
from sympy import Symbol,symbols, solve
import numpy as np
import math

# CÔNG THỨC BAYES. CÁC LUẬT PHÂN BỐ XÁC SUẤT
**XÁC SUẤT TRONG KHOA HỌC DỮ LIỆU**

## Bài 1. 
Một cặp trẻ sinh đôi có thể do cùng trứng (gọi là sinh đôi thật) hay khác trứng (gọi là
sinh đôi giả). Các cặp thật luôn có cùng giới tính. Cặp sinh đôi giả thì giới tính của mỗi bé
độc lập với nhau và có xác suất $0.5$ là con trai, $0.5$ là con gái. Thống kê cho thấy $34%$ cặp
sinh đôi đều là trai, $30%$ cặp sinh đôi đều là gái và $36%$ cặp sinh đôi có giới tính khác nhau.
1) Tìm tỷ lệ cặp thật trên tổng số các cặp sinh đôi. <br>
2) Tìm tỷ lệ cặp thật trên tổng số cặp sinh đôi cùng giới tính.

## Lời giải

Đặt các biến cố:
* $A$ là "cặp sinh đôi cùng trứng (sinh đôi thật)"
* $B$ là "cặp sinh đôi cùng giới tính"

a) Gọi $x$ là xác suất có một cặp sinh đôi thật. Khi đó $P(A)=x$ và $P(A^c) = 1-x$. <br>
* Trong cặp sinh đôi thật: xác suất cùng giới là $1$, xác suất khác giới là $0$.
* Trong cặp sinh đôi giả: xác suất cùng giới là $\frac{1}{2}$, xác suất khác giới là $\frac{1}{2}$.

Theo công thức xác suất đầy đủ thì $$P(B^c) = P(A).P(B^c|A) + P(A^c).P(B^c|A^c) = x.0 + (1-x).\frac{1}{2}.$$

Tổng số cặp sinh đôi là: 

In [2]:
34+30+36

100

Xác suất sinh đôi khác giới tính là:

In [3]:
P_Bc=F(36,100)
print(P_Bc)

9/25


In [4]:
x = symbols('x')
sol = solve((1-x)*(1/2)-P_Bc)
sol

[0.280000000000000]

In [5]:
x= Symbol('x')
solve([(1-x)*(1/2)-P_Bc])

{x: 0.280000000000000}

Sau khi giải phương trình, ta tìm được xác suất sinh đôi thật là $P(A) = x = 0.28.$

b) Tỷ lệ cặp sinh đôi cùng giới tính là:

In [6]:
P_B = (34+30)/100
P_B

0.64

Tỷ lệ cần tìm (sinh đôi thật trên tỷ lệ cùng giới tính): $$P(A|B) = \frac{P(AB)}{P(B)}=\frac{P(A)P(B|A)}{P(B)} = \frac{0.28}{0.64}$$

In [7]:
P = 0.28/0.64
print(f"{P:.2f}")

0.44


Vậy xác suất cần tìm là $0.44$.

## Bài 2. 
Trong số bệnh nhân ở một bệnh viện có $50%$ điều trị bệnh A, $30%$ điều trị bệnh $B$ và
$20%$ điều trị bệnh C. Xác suất để chữa khỏi các bệnh A, B và C tương ứng là $0.7$; $0.8$ và $0.9.$
Tính tỷ lệ bệnh nhân được chữa khỏi bệnh A trong tổng số bệnh nhân đã được chữa khỏi.

Ta có Tỷ lệ bệnh nhân $A, B, C$ là: $50\%, 30\%, 20\%.$ <br>
Xác suất để khỏi bệnh $A, B, C$ là: $0.7, 0.8, 0.9.$

Gọi $E$ là biến cố "khỏi bệnh":
$$P(E) = P(A).P(E|A) + P(B).P(E|B) + P(C).P(E|C)$$
 $$= 0.5 * 0.7 + 0.3 * 0.8 + 0.2 * 0.9 $$

In [8]:
P_E = 0.5 * 0.7 + 0.3 * 0.8 + 0.2 * 0.9
P_E

0.77

Xác suất của biến cố khỏi bệnh $P(E) = 0.77.$

Tính tỷ lệ bệnh nhân được chữa khỏi bệnh A trong tổng số bệnh nhân đã được chữa khỏi
$$P(A|E) = \frac{P(AE)}{P(E)}=\frac{P(A)P(E|A))}{P(E)} = \frac{0.5*0.7}{P(E)}.$$

In [9]:
P = 0.5*0.7  / P_E
print(f"{P:.2f}")

0.45


Xác suất cần tìm là $P(A|E) = 0.45.$

## Bài 3:

Kết quả của một xét nghiệm bệnh sốt xuất huyết trên trẻ em tại một cơ sở y tế được ghi nhận theo bảng sau (trong đó a b c d , , ,là các số nguyên dương):

|    **Kết quả xét nghiệm**     | **Thực tế có bị bệnh**  | **Thực tế không bị bệnh**  | 
|-------------------------------|-------------------------|----------------------------|
| **Dương tính**                | **a**                   |   **b**                    | 
| **Âm tính**                   | **c**                  |     **d**                  | 

Cho biết rằng có hai tiêu chí để đánh giá độ chính xác của dữ liệu là: “độ nhạy” (xét nghiệm 
đúng người bị bệnh) và “độ đặc hiệu” (xét nghiệm đúng người không bị bệnh) và chúng 
được tính bởi: 
$$ \textrm{Độ nhạy} = \frac{a}{a+c};     \textrm{Độ đặc hiệu} = \frac{d}{b+d}$$
<br>
Giả sử xét nghiệm trên có độ nhạy là 95% và độ đặc hiệu là 99%. Biết rằng tỷ lệ bệnh sốt 
xuất huyết trên trẻ em được khảo sát là 5%. Hỏi trong những người có kết quả xét nghiệm 
dương tính, có bao nhiêu phần trăm thật sự mắc bệnh sốt xuất huyết


Giải: Ta có $\textrm{Độ nhạy} = \frac{a}{a+c} = 0,95 \Leftrightarrow a+c=\frac{a}{0.95}; \textrm{Độ đặc hiệu} = \frac{d}{b+d}=0.99\Leftrightarrow b+d=\frac{d}{0.99} \Rightarrow b=\frac{d}{0.99}-d $<br>
a+b+c+d = 100% mà tỷ lệ bệnh sốt xuất huyết trên trẻ em được khảo sát là 5% $\Rightarrow$ $a=0.05(a+b+c+d)$<br>



 mà: $$a=0.05(a+b+c+d)$$
 $$\Leftrightarrow a=0.05(\frac{a}{0.95}+\frac{d}{0.99})$$<br>
 $$\Leftrightarrow a = \frac{0.0495}{0.9405}a+\frac{0.0475}{0.9405}d$$<br>
  $$\Leftrightarrow a = \frac{1}{19}a+\frac{5}{99}d$$<br>
    $$\Leftrightarrow \frac{18}{19}a=\frac{5}{99}d$$.<br>
      $$\Leftrightarrow a=\frac{95}{1782}d$$

Tổng những người có kết quả dương tính là a+b
vậy xác suất để người thật sự mắc bệnh là $$P = \frac{a}{a+b}$$<br>
            $$\Leftrightarrow P=\frac{\frac{95}{1782}d}{\frac{95}{1782}d+(\frac{d}{0.99}-d)}$$

In [53]:
a=95/1782
b=1/0.99-1
P= a/(a+b)
P

0.840707964601769

Vậy tỷ lệ để người dương tính thực sự mắc bệnh là : $P=84,1$%

## Bài 4.

Cho bảng phân phối xác suất sau đây ứng với biến ngẫu nhiên X:

|    **X**     | -3  | -2  | -1  | 0  | 1   | 2   | 3   | 4   |
|--------------|-----|-----|-----|----|-----|-----|-----|-----|
| **P(X=xi)**   | 0.2 | 0.15| 0.05| 0.1| 0.1 |0.05 | 0.05| **a** |



a) Xác định giá trị của a, từ đó tìm kỳ vọng, phương sai của $ X$.<br>
b) Cho $Y = 4 – 5X$ . Không lập bảng phân bố, hãy tìm cách tính kỳ vọng và phương sai của $ Y$.

Tổng các giá trị của $P(X=xi) = 1$

Vậy $a = 1- \sum_{i}P(X=xi)$

In [10]:
a=1-(0.2+0.15+0.05+0.1+0.1+0.05+0.05)
round(a,2)

0.3

$ a=0.3$

In [11]:
PX_xi =np.array([0.2 , 0.15, 0.05, 0.1 , 0.1 , 0.05, 0.05, 0.3 ])
PX_xi

array([0.2 , 0.15, 0.05, 0.1 , 0.1 , 0.05, 0.05, 0.3 ])

In [12]:
X=np.array([-3,-2,-1,0,1,2,3,4])
X

array([-3, -2, -1,  0,  1,  2,  3,  4])

Kì vọng là : $\sum_{i}^{} P(X=xi)* xi$

In [13]:
E=PX_xi*X
print(E)
E_x=E.sum()
round(E_x,2)

[-0.6  -0.3  -0.05  0.    0.1   0.1   0.15  1.2 ]


0.6

Vậy :$E[x]=0.6$

**Công thức tính phương sai từ Kỳ vọng** : $$Var[x]= E[x^2] - (E[x])^2$$

In [14]:
E_sq=PX_xi*(X*X)
E_sq.sum()

8.0

In [15]:
round(pow(E_x,2),4)

0.36

$E[x^2]=8,(E[x])^2=0.36$

In [16]:
var_x=E_sq.sum()-E_x**2
round(var_x,3)

7.64

Vậy: $Var[x]=7.64$

In [17]:
std_x=math.sqrt(var_x)
round(std_x,4)

2.7641

Vậy $std[x]=2.7641$

b)Cho $ 𝑌=4–5𝑋 $ . Không lập bảng phân bố, hãy tìm cách tính kỳ vọng và phương sai của  $𝑌 $
<br>
**Áp dụng công thức**:
$$E(aX + b) =aE(X) +b $$
$$ Var(aX + b) = a^2Var(X).$$

In [18]:
E_y=-5*E_x+4
round(E_y,4)

1.0

**Kỳ vọng của y**: $E[y]=1$

$$Var[x]= E[x^2] - (E[x])^2$$

In [19]:
var_y = 5**2*var_x
var_y

191.0

**Phương sai của y:** $Var[y]=191$

In [20]:
Std_y=np.sqrt(var_y)
Std_y

13.820274961085254

## Bài 5

Hai xạ thủ A và B tập bắn, mỗi người bắn hai phát. Xác suất bắn trúng đích của A trong 
mỗi phát là $0.4$ còn của B là $0.5$. Gọi $X$ là số phát trúng của A trừ đi số phát trúng của B. Tìm 
phân bố xác suất của $X$, tìm kỳ vọng, phương sai.

 Gọi $A_yB_z$ với y,z lần lượt là số lần trúng đạn của A và B:<br>
 Các trường hợp có thể xảy ra:
 $$(A_yB_z,p(A_yB_z),X):(A_2B_0=0.04=2, A_1B_0=0.12 =1, A_0B_0=0.09=0, A_2B_1=0.08=1, A_1B_1=0.24=0, A_0B_1=0.18=-1, A_2B_2=0.04=0, A_1B_2=0.12=-1, A_0B_2=0.09=-2)$$<br>
 $P(X=-2) = A_0B_2 = 0.09$<br>
 $P(X=-1) = A_0B_1+A_1B_2 = 0.18+0.12=0.3$<br>
 $P(X= 0)= A_0B_0+A_1B_1+A_2B_2= 0.09+0.24+0.04=0.37$<br>
 $P(X= 1)= A_1B_0+ A_2B_1 =0.12+0.08=0.2$<br>
 $P(X= 2) = A_2B_0 = 0.04$ <br>

Ta có bảng phân phối xác suất như sau :

|    **x**     | -2  | -1  | 0  | 1  | 2   |
|--------------|-----|-----|----|----|-----|
| **P(X=xi)**  | 0.09 |0.3| 0.37| 0.2| 0.04|

Ta có :Kì vọng là : $\sum_{i}^{} P(X=xi)* xi$

In [21]:
PX_xi= np.array([0.09,0.3,0.37,0.2,0.04])
x=np.array([-2,-1,0,1,2],dtype=float)

In [22]:
E_x=sum(PX_xi*x)
round(E_x,4)

-0.2

**Kỳ vọng**: $E[x]=-0.2$

$$Var[x]= E[x^2] - (E[x])^2$$

In [23]:
E_xx=sum(PX_xi*(x*x))
round(E_xx,4)

1.02

$E[x^2] = 1.02$

**Phương sai**: $Var[x]=1.02-(-0.2)^2$

In [24]:
Var_x= E_xx-pow(E_x,2)
round(Var_x,3)

0.98

Vậy $Var[x] = 0.98$

## Bài 6

Trong một thành phố nhỏ, trung bình một tuần có 2 em bé sinh ra. Tính xác suất để:<br>
a) Không có em nào sinh ra trong vòng một ngày.<br>
b) Có ít nhất 3 em bé sinh ra trong vòng hai ngày.

**Áp Dụng Phân phối Poisson** $$P(X=xi) = e^{-\lambda} \frac{\lambda^x}{x!}$$

**a) Vì 1 tuần có 2 em bé sinh ra nên 1 ngày sẽ có 2/7 em bé được sinh ra nên $\lambda_1 = 2/7$**

In [25]:
lambda1=2/7

gọi $P(X_0)$ là xác suất không có em bé nào sinh ra trong 1 ngày <br>
áp dụng phân phối poisson ta được <br>
$P(X=xi) = e^{-\lambda} \frac{\lambda^0}{0!}$

In [26]:
PX_0 = pow(math.e,-lambda1)*(pow(lambda1,0)/math.factorial(0))
round(PX_0,4)

0.7515

Vậy xác suất **không có em bé nào sinh ra trong 1 ngày** là : $P(X_0)$=0.7515

**b) $\lambda$ của 2 ngày sẽ $\lambda_1 *2 \rightarrow \lambda_2 = 4/7$**

In [27]:
lambda2=F(4,7)
print(lambda2)

4/7


Ta thấy $P(X>=3)= 1-(P(X=0)+P(X=1)+P(X=2))$<br>
$ = 1 - e^{-\lambda} \frac{\lambda^0}{0!}- e^{-\lambda} \frac{\lambda^1}{1!}- e^{-\lambda} \frac{\lambda^2}{2!}$

gọi $P(X_2)$ là xác suất có ít nhất 3 em bé sinh ra trong vòng 2 ngày 
  

In [28]:
PX_2= 1-pow(math.e,-lambda2)*(pow(lambda2,0)/math.factorial(0)) -pow(math.e,-lambda2)*(pow(lambda2,1)/math.factorial(1)) -pow(math.e,-lambda2)*(pow(lambda2,2)/math.factorial(2))
round(PX_2,4)

0.0204

Vậy xác suất để có ít nhất **3 em bé sinh ra trong vòng 2 ngày** là: $P(X_2)=0.0204$

## Bài 7

Một bài thi trắc nghiệm có 10 câu hỏi, mỗi câu có 5 phương án chọn và chỉ có một <br>
phương án là đúng. Mỗi câu đúng được 1 điểm, sai bị trừ 0.5 điểm. Một SV vào đánh bừa hết cả 10 câu.<br> 
a) Hãy tính xác suất để SV được đúng 8đ và xác suất để SV bị điểm âm.<br>
b) Giả sử mỗi lần bị điểm âm, SV sẽ thi tiếp đến khi nào được điểm không âm thì thôi. Tính 
xác suất để SV dừng thi ở lần thứ 4

In [29]:
for i in range(0,11):
    print('Số câu đúng: '+str(i)+', Điểm: '+str(i-(10-i)*0.5))

Số câu đúng: 0, Điểm: -5.0
Số câu đúng: 1, Điểm: -3.5
Số câu đúng: 2, Điểm: -2.0
Số câu đúng: 3, Điểm: -0.5
Số câu đúng: 4, Điểm: 1.0
Số câu đúng: 5, Điểm: 2.5
Số câu đúng: 6, Điểm: 4.0
Số câu đúng: 7, Điểm: 5.5
Số câu đúng: 8, Điểm: 7.0
Số câu đúng: 9, Điểm: 8.5
Số câu đúng: 10, Điểm: 10.0


Nếu xét trường hợp thì sẽ không bao giờ có trường hợp SV được đúng 8 điểm <br>
Vậy nên bài này em sẽ tính lại thành xác suất để SV đúng được 8 câu

Áp dụng Phân phối nhị thức $$f (x) = P(X = x) = C_n^xp^x(1-p)^{n-x}$$

**a1)Gọi $P(c_8)$ là xác suất SV đúng được 8 câu**<br>
a là xác suất sinh viên làm đúng 1 câu $\rightarrow$ a=1/5

In [30]:
a=F(1,5)
print(a)

1/5


In [31]:
P_8= math.comb(10,8)*pow(a,8)*pow((1-a),(10-8))
print(P_8)
print(round(P_8,7))

144/1953125
737/10000000


In [32]:
round(144/1953125,4)

0.0001

**Vậy xác suất để sinh viên làm đúng 8 câu là :**$P(c_8)=144/1953125$

a2) Gọi $P(mn)$ là xác suất để SV được điểm âm <br>
ta có số trường hợp để SV được điểm âm lần lượt làm đúng từ [0,3] câu
vậy: $$P(mn) = \sum_{i=0}^{3} C_n^ip^i(1-p)^{n-i}$$

In [33]:
1/5

0.2

In [34]:
P_mn=0
n=10
for i in range(0,4):
    P_mn+= math.comb(n,i)*pow(a,i)*pow((1-a),(n-i))
print(P_mn)

8585216/9765625


In [35]:
round(8585216/9765625,4)

0.8791

vậy xác suất để SV được điểm âm là : $P(mn) \approx 0.8791 $

**b)** Áp dụng phân phối hình học : $$P(X=x) = (1-p)^{x-1}p$$

In [36]:
round(1180409/9765625,4)

0.1209

Gọi xác suất SV được điểm dương $P(nm) = P(\bar{m}\bar{n}) =1-P(mn) = 1-0,8791=0.1209$

In [37]:
P_nm=0.1209

gọi $P(B)$ là xác suất mà sinh viên thi lần thứ 4 được điểm dương

$$P(X=x) = (1-p)^{x-1}p$$

In [38]:
P_B= pow((1-P_nm),4-1)*P_nm
print(round(P_B,4))

0.0821


Vậy xác suất để sinh viên dừng thi ở **lần thứ 4 là** $P(B)=0.0821$